### PR TITLE
[SDFAB-1202] Prototyping Host discovery functionality

### PR DESCRIPTION
**Pull request by eroshiva**

### CODE REVIEW

File: pkg/linkdiscovery/controller.go [hunk 1]
```diff
@@ -197,6 +211,36 @@ func (c *Controller) setStateIf(condition State, state State) {
...
+		c.hosts[macString] = host
+		log.Infof("Added a new host: %s <- %s/%s", macString, ipString, port)
+	}
+}
+
+// FIXME: rework this function
+func (c *Controller) pruneHosts() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	limit := time.Now().Add(-30 * time.Minute) // this is to discuss
```
Comment (tomikazi) [inline pkg/linkdiscovery/controller.go:235]: Yep, we'll probably make this configurable via gNMI.

File: pkg/linkdiscovery/controller.go [hunk 2]
```diff
@@ -242,6 +286,7 @@ func (c *Controller) enterLinkDiscovery() {
 	tPorts := time.NewTicker(time.Duration(c.config.PortRediscoveryFrequency) * time.Second)
 	tPrune := time.NewTicker(time.Duration(c.config.LinkPruneFrequency) * time.Second)
 
+	// Do I have to emit ARP packets here? I guess so...
```
Comment (tomikazi) [inline pkg/linkdiscovery/controller.go:289]: I don't think so, unless we want to provoke the hosts to reply.

File: pkg/linkdiscovery/connection.go [hunk 3]
```diff
@@ -171,6 +172,16 @@ func (c *Controller) processPacket(packetIn *p4api.PacketIn) {
 		}
 		c.updateIngressLink(pim.IngressPort, uint32(egressPort), string(lldp.ChassisID.ID))
 	}
+
+	// if condition to process ARP packet
+	arpLayer := rawPacket.Layer(layers.LayerTypeARP)
+	if arpLayer != nil {
+		pim := c.codec.DecodePacketInMetadata(packetIn.Metadata)
+		arp := arpLayer.(*layers.ARP)
+		// FIXME: pim.RoleAgentID substitutes device.ID - is it correct? Should I abandon it?
```
Comment (tomikazi) [inline pkg/linkdiscovery/connection.go]: I would not worry about this here.  We don't need this.

File: pkg/linkdiscovery/connection.go [hunk 4]
```diff
@@ -171,6 +172,16 @@ func (c *Controller) processPacket(packetIn *p4api.PacketIn) {
...
+		c.updateHost(packet.MACString(arp.SourceHwAddress), packet.IPString(arp.SourceProtAddress),
+			fmt.Sprintf("%d/%d", pim.RoleAgentID, pim.IngressPort))
```
Comment (tomikazi) [inline pkg/linkdiscovery/connection.go]: I would track the `uint32` IngressPort only.

**Review by tomikazi — COMMENTED** [review]
Just a few minor things. Looks perfectly on track. :)

File: pkg/linkdiscovery/controller.go (hunk 1 above)
Comment (eroshiva) [reply pkg/linkdiscovery/controller.go:235]: Alright, left unchanged per our discussion.

File: pkg/linkdiscovery/controller.go (hunk 2 above)
Comment (eroshiva) [reply pkg/linkdiscovery/controller.go:289]: Left out

File: pkg/linkdiscovery/connection.go (hunk 4 above)
Comment (eroshiva) [reply pkg/linkdiscovery/connection.go]: Addressed

File: pkg/linkdiscovery/connection.go (hunk 3 above)
Comment (eroshiva) [reply pkg/linkdiscovery/connection.go]: Comment was removed

File: pkg/linkdiscovery/connection.go [hunk 5]
```diff
@@ -220,10 +234,13 @@ func (c *Controller) emitLLDPPackets() {
 	log.Info("LLDP packets emitted")
 }
 
-func (c *Controller) installPuntRule(tableID uint32, actionID uint32, ethTypeMatchFieldID uint32, setRoleAgentParamID uint32) error {
+func (c *Controller) installPuntRule(tableID uint32, actionID uint32, ethTypeMatchFieldID uint32, setRoleAgentParamID uint32, arp bool) error {
```
Comment (tomikazi) [inline pkg/linkdiscovery/connection.go]: I would just pass in the ethType.

File: pkg/linkdiscovery/connection.go [hunk 6]
```diff
@@ -220,10 +234,13 @@ func (c *Controller) emitLLDPPackets() {
...
 	ethTypeValue := []byte{0, 0}
-	binary.BigEndian.PutUint16(ethTypeValue, uint16(layers.EthernetTypeLinkLayerDiscovery))
-
+	if arp {
```
Comment (tomikazi) [inline pkg/linkdiscovery/connection.go]: See above

**Review by tomikazi — COMMENTED** [review]
Almost there... can you also rename the `linkdiscovery` package to just `discovery` while we're doing this?

File: pkg/linkdiscovery/connection.go (hunk 5 above)
Comment (eroshiva) [reply pkg/linkdiscovery/connection.go]: Addressed

File: pkg/linkdiscovery/connection.go (hunk 6 above)
Comment (eroshiva) [reply pkg/linkdiscovery/connection.go]: Addressed

**Review by tomikazi — APPROVED** [review]

**Pull request merged**